Support returning max allowed current at grid connection point

tiyash-basu-frequenz opened the pull request:
This PR
* adds component-category-specific metadata, and
* introduces metadata for the grid component category, and
* adds the max allowed current to the grid metadata

tiyash-basu-frequenz committed:
Add grid-connection-point metadata

This commit introduces a `grid.Metadata` message, which contains the item
`rated_fuse_current`. This is the rating of the fuse at the grid connection
point.
This rating specifies the maximum amount of current, measured in amperes,
that can flow in or out of each of the 3 phases individually.
The current _i_ A at the grid connection point must comply with the
following constraint: : `-rated_fuse_current <= i <= rated_fuse_current`

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz committed:
Add metadata specific to component category

This metadata is returned in response to `ListComponents` calls,
in a new message variable `Component.metadata`.

This is a more general way of representing category-specific metadata,
like type, and removes `Component.type`.

Signed-off-by: Tiyash Basu <tiyash.basu@frequenz.com>

tiyash-basu-frequenz commented:
An alternative would have been to just add a `max_current_magnitude` item to the `Component` message, but that would have introduced two different methods of setting bounds on inverters and batteries (the other one being dynamic bounds). Would be nice to hear your thoughts on this idea as well, besides the PR.

thomas-nicolai-frequenz commented on proto/frequenz/api/microgrid/grid.proto:
```diff
@@ -0,0 +1,19 @@
+  // The maximum current magnitude in amperes.
+  // This is the maximum current that can be drawn from or supplied to
+  // the grid.
+  uint32 max_current_magnitude = 1;
```
Magnitude? I've never experienced us using this term so far. Why couldn't it just be name `max_current` instead?

thomas-nicolai-frequenz replied:
If we wanna introduce this term here we need to describe it better in the comment what that is.

tiyash-basu-frequenz replied:
The maximum current at a fuse is a signed value. Meaning, if the fuse limit is 5A, then a current _i_ A should be `-5 <= i <= 5`, or `|i| <= 5`. Hence I added the term `magnitude`, to indicate that this limit applies to +ve as well as -ve currents.

Or do we want to split this? Then we could specify that at a site we can consume from the grid, but never supply to it.

tiyash-basu-frequenz replied:
>Or do we want to split this? Then we could specify that at a site we can consume from the grid, but never supply to it.

I made the changes.

thomas-nicolai-frequenz replied:
> Hence I added the term magnitude, to indicate that this limit applies to +ve as well as -ve currents.

Well thats where the inconsistency comes in. We have signed values in many places and yet don't call them all `magnitudes`. I think its enough to have that in the description tbh otherwise we would need to change this all over the place.

thomas-nicolai-frequenz replied:
> I made the changes

Don't! This is the fuse limit. A fuse imho doesn't care in which direction the power is flowing. This I would call this `max_grid_fuse_current` or something tbh. I would suggest to later introduce virtual limits for grid feed-in or grid consumption once we need that. More likely we will add frequency limits later on as the first step but lets postpone that discussion until we really need it.

tiyash-basu-frequenz replied:
Done. I renamed the variable to `max_grid_fuse_current`.

sahas-subramanian-frequenz replied:
is this checked in?  I still see 3 vars, one for each phase.

tiyash-basu-frequenz replied:
Not yet. I am waiting for further input from Niklas. I will update the PR once I have all the details.

tiyash-basu-frequenz replied:
Renamed it to `rated_fuse_current`.

https://github.com/frequenz-floss/frequenz-api-microgrid/pull/38#issuecomment-1514732619

leandro-lucarella-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -28,6 +28,12 @@
 ## Bug Fixes
 
-<!-- Here goes notable bug fixes that are worth a special mention or explanation -->
+<!-- Here goes notable bug fixes that are worth a special mention or explanation -->`
```
I guess this change is not intended?

tiyash-basu-frequenz replied:
oops, looks like a typo. Will fix.

tiyash-basu-frequenz replied:
Fixed.

leandro-lucarella-frequenz commented on proto/frequenz/api/microgrid/microgrid.proto:
```diff
@@ -412,15 +413,14 @@ message Component {
-    sensor.Type sensor = 10;
+  // The metadata specific to the component category.
+  oneof metadata {
+    grid.Metadata grid = 10;
```
Shouldn't this start at 11 to avoid repeating used tags?

Maybe also add a comment here saying which tags were used in the past and should not be reused?

tiyash-basu-frequenz replied:
> Shouldn't this start at 11 to avoid repeating used tags?

I don't think repeating used tags is a concern, because this is going to be rolled out as a breaking change. But if you really want new tags, I can update these.

leandro-lucarella-frequenz replied:
OK, it was strange that you use new tags for all the metadata except for the new `grid` one, so it looked like something not intentional.

In any way, IIRC, it is a good practice to not reuse tags in protobufs, as at the protocol level if we do if for some reason we have mismatched versions between the server and the client, and the server is sending some `grid` metadata the client might interpret it as some old `sensor` type instead, maybe making the client code blow up in a more unexpected way.

I agree with such big change it will probably blow up pretty soon and consistently but if the changes were smaller it could be a harder to find issue if versions are accidentally mixed.

In any case, I'd strongly prefer to either:
* Not reuse tags
* Keep tags continuous

Right now it is weird that we suddenly jump from 5 to 10 for no apparent reason (if you don't see this PR :).

tiyash-basu-frequenz replied:
>so it looked like something not intentional.

I just did it to keep `10` available for `grid` in the 2nd commit :)

> at the protocol level if we do if for some reason we have mismatched versions between the server and the client, and the server is sending some grid metadata the client might interpret it as some old sensor type instead, maybe making the client code blow up in a more unexpected way.

Yes, that will happen, and I agree that we should not reuse tags (in fact it's a standard practice to not reuse tags).
But we are at a very early stage, and doing it now will blow up the tag numbers very quickly. I'd suggest waiting until we release v1.0.0 until we introduce such constraints. 

> Right now it is weird that we suddenly jump from 5 to 10 for no apparent reason (if you don't see this PR :).

True. We have left intentional gaps in places so that we can introduce new items in a non-breaking change.
But this is not required, since the numbers can be arbitrary. We can fix those in a cleanup PR.

tiyash-basu-frequenz replied:
I'll just go ahead an renumber the tags, so that older ones are not used.

... and done.

leandro-lucarella-frequenz replied:
OK, as I said before, for me consistency/information is the most important. If you want to reuse tags because it is very early in the development, then sure, but I would clarify in the commit message for example, and chose contiguous tags.

If there is a gap in the tags, either add a comment in the code saying the gap is there because there were some old attributes using those tags so they shouldn't be reused, or maybe have it as a general rule of the project that gaps can't be reused (but IMHO being explicit locally is always better and less error prone).

tiyash-basu-frequenz replied:
Fair enough 👍

leandro-lucarella-frequenz reviewed (commented):
Other than the few comments, LGTM.

leandro-lucarella-frequenz commented on proto/frequenz/api/microgrid/grid.proto:
```diff
@@ -0,0 +1,22 @@
+  // into or out of the grid connection point. E.g., a current _i_ A at the grid
+  // connection point has to follow the constraint:
+  // -max_grid_fuse_current <= i <= max_grid_fuse_current
+  uint32 max_grid_fuse_current = 1;
```
Also minimal, but since this is already in the scope of the grid component category maybe we can just leave it at `max_fuse_current`.

tiyash-basu-frequenz replied:
Done

tiyash-basu-frequenz commented:
Updated the docs to specify that the limit applies individually to all the 3 phases.

leandro-lucarella-frequenz reviewed (dismissed):
As said in the previous comment, I would add a comment about the gap in the code for extra clarity, but not a blocker, so approving.

tiyash-basu-frequenz commented:
Following our discussion [here](https://github.com/frequenz-floss/frequenz-sdk-python/discussions/322#discussioncomment-5603799), I have updated the PR to introduce a `Fuse` message with limits for individual phases, and added it to the `grid.Metadata` message

tiyash-basu-frequenz commented:
I reverted the last change according to the latest discussions and inputs.

Now we just have the following:
```proto
// The grid connection point metadata.
message Metadata {
  // The rating of the fuse at the grid connection point.
  // This rating specifies the maximum amount of current, measured in amperes,
  // that can flow in or out of each of the 3 phases individually.
  // The current _i_ A at the grid connection point must comply with the
  // following constraint: : `-max_fuse_current <= i <= max_fuse_current`
  uint32 max_fuse_current = 1;
}

```

thomas-nicolai-frequenz commented:
> max_fuse_current

Shall we call it `fuse_current_limit` instead? the term max feels a little ambiguous.

leandro-lucarella-frequenz commented:
`fuse_limit_current`? So we can change the suffix if there are new limits.

It is quite unfortunate that no matter what it is still ambiguous and could be misunderstood as the limit *rigth now* instead of the limit on the electrons flow.

tiyash-basu-frequenz commented:
How about `rated_fuse_current`? The inclusion of the term `fuse` itself implies that the current should be limited here.

thomas-nicolai-frequenz commented:
> It is quite unfortunate that no matter what it is still ambiguous and could be misunderstood as the limit rigth now instead of the limit on the electrons flow.

true.

tiyash-basu-frequenz commented:
>How about `rated_fuse_current`? The inclusion of the term fuse itself implies that the current should be limited here. 

Done. Ready for another look.

leandro-lucarella-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -28,6 +28,21 @@
+  This is a more general way of representing category-specific metadata,
+  like category-type, and deprecates `Component.type`.
+
+* [Introduced grid max-current](https://github.com/frequenz-floss/frequenz-api-microgrid/pull/38).
+  This change introduces a `grid.Metadata` message, which contains the item
+  `max_fuse_current`. This is the rating of the fuse at the grid connection
+  point.
+  This rating specifies the maximum amount of current, measured in amperes,
+  that can flow in or out of each of the 3 phases individually.
+  The current _i_ A at the grid connection point must comply with the
+  following constraint: : `-max_fuse_current <= i <= max_fuse_current`
```
This will need to be updated with the new name.

tiyash-basu-frequenz replied:
Fixed 🚀

leandro-lucarella-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -28,6 +28,21 @@
+  This metadata is returned in response to `ListComponents` calls,
+  in a new message variable `Component.metadata`.
+  This is a more general way of representing category-specific metadata,
+  like category-type, and deprecates `Component.type`.
```
This is actually **removing**, not deprecating.

tiyash-basu-frequenz replied:
Fixed 🚀

leandro-lucarella-frequenz reviewed (commented):
Other than a few comments about outdated docs, it looks like we have a winner!

leandro-lucarella-frequenz reviewed (approved):
Unless there is any pending topic in the SDK discussion that need to be considered here, this is, as far as I am concerned, ready to go! :shipit:

tiyash-basu-frequenz commented:
@thomas-nicolai-frequenz Do you have any further input on this? If not, we can get this merged.

thomas-nicolai-frequenz commented:
I think I have created enough trouble :-) Lets merge it!
